### PR TITLE
Loom 1.3: Legacy Forge

### PR DESCRIPTION
**Pull request by Jab125**
TODO list:
- [x] Generate forge sources
- [x] `dev.architectury.loom.metadata.McModInfo`
- [x] Workaround No-Intermediary remapping bug (currently it is using legacy fabric intermediary)
- [x] Forge's access transformer
- [x] Mod's access transformer
- [x] Project's access transformer
- [x] Add project's access transformer(s!) to `META-INF/MANIFEST.MF`
- [x] Access widener -> Access transformer conversion
- [x] Confirm coremod functionality
- [x] Fix `ObfuscationReflectionHelper` on `1.12.2-14.23.5.2816` by fixing `FMLDeobfuscatingRemapper`
- [x] Fix forge not finding log4j-core (fixed by adding `org.apache.logging.log4j:log4j-core:2.11.2` to the classpath)
- [x] Use a better LZMA reader
- [x] Find a pack200 reader that doesn't have a sticky license
- [x] Remove `Pack200Provider`, obsolete.
- [x] Patch forge's obfuscation detection to not use a hardcoded name.
- [ ] Fix legacy yarn mappings not applying properly (catching the errors reveals that the names are matched incorrectly)
- [ ] Allow mapping class names from SRG to named in forge's own class remapper
- [ ] Fix all hardcoded instances of hardcoded SRG names in forge
- [x] Integration Tests

### CODE REVIEW

**Comment by wagyourtail** [comment]
> Find a pack200 reader that doesn't have a sticky license

 `org.apache.commons:commons-compress`
 Practically a drop in replacement
Maybe will also work for "better lzma reader" as well

**Comment by Jab125** [comment]
> > Find a pack200 reader that doesn't have a sticky license
> 
> `org.apache.commons:commons-compress` Practically a drop in replacement Maybe will also work for "better lzma reader" as well

done

File: src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/AccessTransformSetMapper.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.configuration.providers.mappings.utils;
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/mappings/utils/AccessTransformSetMapper.java]: Should be in `dev.architectury.loom.util` along with any new util classes

File: src/main/java/net/fabricmc/loom/util/ModPlatform.java [hunk 2]
```diff
@@ -48,21 +53,50 @@ public boolean isExperimental() {
...
+
+			for (String name : names) {
+				String lowercaseName = name.toLowerCase(Locale.ROOT);
+				platformList.append(lowercaseName).append(", ");
+				loomPlatform.append("['loom.platform = ").append(lowercaseName).append("'], ");
+			}
+
+			platformList.setLength(platformList.length()-2);
+			platformList.append("]");
+			loomPlatform.setLength(loomPlatform.length()-2);
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/util/ModPlatform.java]: You can use `StringJoiner` to have less convoluted code here.

File: src/main/java/net/fabricmc/loom/util/ModPlatform.java [hunk 3]
```diff
@@ -48,21 +53,50 @@ public boolean isExperimental() {
...
 
-	public static void assertPlatform(LoomGradleExtensionAPI extension, ModPlatform platform) {
-		assertPlatform(extension, platform, () -> {
-			String msg = "Loom is not running on %s.%nYou can switch to it by adding 'loom.platform = %s' to your gradle.properties";
-			String name = platform.name().toLowerCase(Locale.ROOT);
-			return msg.formatted(name, name);
-		});
+	public static void assertPlatform(LoomGradleExtensionAPI extension, ModPlatform... platforms) {
+		assertPlatform(extension, () -> {
+			String msg = "Loom is not running on any of %s.%nYou can switch to it by any of the following: Add any of %s to your gradle.properties";
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/util/ModPlatform.java:72]: Nit: This should not say "any of [FORGE]" for single-platform cases. Probably best to just check if `platforms.length == 1` and use the old message format.

**Review by Juuxel — COMMENTED** [review]
Some notes, nothing like a full review yet

**Comment by Jab125** [comment]
95% working, going to submit for review

**Comment by Jab125** [comment]
Will redo later for Loom 1.4, since that adds `isForgeLike()`

**Pull request closed without merging**